src/sources/pulsar.rs

```diff
@@ -0,0 +1,451 @@
+    #[derivative(Default(value = "default_framing_message_based()"))]
+    framing: FramingConfig,
+
+    #[configurable(derived)]
+    #[serde(default = "default_decoding")]
+    #[derivative(Default(value = "default_decoding()"))]
+    decoding: DeserializerConfig,
+
+    #[configurable(derived)]
+    #[serde(default, deserialize_with = "bool_or_struct")]
+    acknowledgements: SourceAcknowledgementsConfig,
+}
+
+/// Authentication configuration.
+#[configurable_component]
+#[derive(Clone, Debug)]
+struct AuthConfig {
+    /// Basic authentication name/username.
+    ///
+    /// This can be used either for basic authentication (username/password) or JWT authentication.
+    /// When used for JWT, the value should be `token`.
+    name: Option<String>,
+
+    /// Basic authentication password/token.
+    ///
+    /// This can be used either for basic authentication (username/password) or JWT authentication.
+    /// When used for JWT, the value should be the signed JWT, in the compact representation.
+    token: Option<SensitiveString>,
+
+    #[configurable(derived)]
+    oauth2: Option<OAuth2Config>,
+}
+
+/// OAuth2-specific authentication configuration.
+#[configurable_component]
+#[derive(Clone, Debug)]
+pub struct OAuth2Config {
+    /// The issuer URL.
+    issuer_url: String,
+
+    /// The credentials URL.
+    ///
+    /// A data URL is also supported.
+    credentials_url: String,
+
+    /// The OAuth2 audience.
+    audience: Option<String>,
+
+    /// The OAuth2 scope.
+    scope: Option<String>,
+}
+
+/// Dead Letter Queue policy configuration.
+#[configurable_component]
+#[derive(Clone, Debug)]
+struct DeadLetterQueuePolicy {
+    /// Maximum number of times that a message will be redelivered before being sent to the dead letter queue.
+    pub max_redeliver_count: usize,
+
+    /// Name of the dead topic where the failing messages will be sent.
+    pub dead_letter_topic: String,
+}
+
+struct FinalizerEntry {
+    consumer: Arc<Mutex<Consumer<String, TokioExecutor>>>,
+    message: pulsar::consumer::Message<std::string::String>,
+}
+
+impl std::fmt::Debug for FinalizerEntry {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("FinalizerEntry")
+            .field("message", &self.message.payload)
+            .finish()
+    }
+}
+
+impl GenerateConfig for PulsarSourceConfig {
+    fn generate_config() -> toml::Value {
+        toml::from_str(
+            r#"
+            topics = ["topic1", "topic2"]
+            endpoint = "pulsar://127.0.0.1:6650""#,
+        )
+        .unwrap()
+    }
+}
+
+#[async_trait::async_trait]
+impl SourceConfig for PulsarSourceConfig {
+    async fn build(&self, cx: SourceContext) -> crate::Result<super::Source> {
+        let consumer = create_consumer(self).await?;
+        let decoder = DecodingConfig::new(
+            self.framing.clone(),
+            self.decoding.clone(),
+            LogNamespace::Legacy,
```

Comment:
We'll want to make sure log namespacing is handled correctly as per this doc: https://github.com/vectordotdev/vector/blob/master/rfcs/2022-04-20-12187-log-namespacing.md

Here is a tutorial: https://github.com/vectordotdev/vector/blob/master/docs/tutorials/lognamespacing.md